### PR TITLE
Editorial: Use named records for DateTimeFormat format records

### PR DESCRIPTION
**Pull request by anba**
This PR aims to restructure [Intl.DateTimeFormat Constructor - Internal Slots](https://tc39.es/ecma402/#sec-intl.datetimeformat-internal-slots) to use named Record types instead of ad-hoc records. In the process of this clean-up, some editorial bugs were found and fixed, see the commit messages for details.

In addition to providing a more _type-safe_ approach when compared to using ad-hoc records, these changes should also help with the Temporal integration, see <https://tc39.es/proposal-temporal/#sec-temporal-intl>.

---

The PR applies on top of #822.

### CODE REVIEW

**Review by ben-allen — APPROVED** [review]
Whew! Reviewed commits one-by-one; all of the comments are about things fixed in later commits. 

Thank you again for all your work on this -- it's awe-inspiring.

File: spec/datetimeformat.html
Comment (anba) [reply spec/datetimeformat.html]: Fixed.

File: spec/datetimeformat.html
Comment (anba) [reply spec/datetimeformat.html]: Fixed.

File: spec/datetimeformat.html
Comment (anba) [reply spec/datetimeformat.html]: Good catch. Fixed to refer to the correct record type in the intermediate commit.

File: spec/datetimeformat.html
Comment (anba) [reply spec/datetimeformat.html]: Correct, the later commit will this to use the correct source.

**Comment by anba** [comment]
Fixed all review comments and rebased the commits to apply cleanly on tip.

**Comment by anba** [comment]
Rebased again.

Do we still wait for another review or can we land these changes?

**Review by sffc — COMMENTED** [review]
@gibson042 If this looks good please merge.

File: spec/datetimeformat.html
Comment (gibson042) [inline spec/datetimeformat.html:384]: I have a goal of replacing each [[rangePatterns]] and [[rangePatterns12]] unit-keyed Record with a List of Records with fields like { [[BiggestDivergentUnit]]: string, [[RangePattern]]: DateTimeFormat Range Pattern Record } Records (or possibly just a list of range pattern Records that include that first field, saving a level of depth), and it might be appropriate to include that change in this PR rather than introduce new bibliographic terms that will just need to be removed. Thoughts?

<details>

The net effect should be to reduce opacity of field semantics, replacing e.g.
```
[[rangePatterns]]: {
  [[Hour]]: {
    [[hour]]: "numeric",
    [[minute]]: "numeric",
    [[PatternParts]]: […]
  },
  [[Minute]]: {
    [[hour]]: "numeric",
    [[minute]]: "numeric",
    [[PatternParts]]: […]
  },
  [[Default]]: {
    [[year]]: "2-digit",
    [[month]]: "numeric",
    …,
    [[PatternParts]]: […]
}
```
with
```
[[rangePatterns]]: [
  { [[BiggestDivergentUnit]]: "hour",
    [[hour]]: "numeric",
    [[minute]]: "numeric",
    [[PatternParts]]: […]
  },
  { [[BiggestDivergentUnit]]: "minute",
    [[hour]]: "numeric",
    [[minute]]: "numeric",
    [[PatternParts]]: […]
  },
  { [[BiggestDivergentUnit]]: ~empty~,
    [[year]]: "2-digit",
    [[month]]: "numeric",
    …,
    [[PatternParts]]: […]
  }
]
```

</details>

File: spec/datetimeformat.html
Comment (gibson042) [inline spec/datetimeformat.html:695]: I'm also thinking about flattening these from [[DateTimeRangeFormat]].[[\<dateStyle>]].[[\<timeStyle>]] to [[DateTimeRangeFormat]].[[\<dateAndTimeStyle>]] with e.g. `Let _dateAndTimeStyle_ be the string-concatenation of _dateStyle_, *"-"*, and _timeStyle_.`, but haven't yet decided if that's worthwhile.

File: spec/locales-currencies-tz.html
Comment (gibson042) [inline spec/locales-currencies-tz.html:423]: :heart_eyes:

File: spec/locales-currencies-tz.html
Comment (gibson042) [inline spec/locales-currencies-tz.html:421]:
```suggestion
      <dfn>Pattern String</dfn> is a String value which contains zero or more substrings of the form *"{key}"*, where key can be any nonempty sequence consisting only of elements from the ASCII word characters. The syntax of the abstract pattern strings is an implementation detail and is not exposed to users of ECMA-402.
```
(trying to avoid use of "name" misleading readers into thinking that e.g. "{0}" and "{1}" are not supported)

**Review by gibson042 — COMMENTED** [review]
This is great stuff, but see my comments about surfacing "pattern" in the new names.

**Comment by ben-allen** [comment]
@anba I've rebased this onto the current main on my own fork:  https://github.com/ben-allen/ecma402/tree/anba-826-working-branch. If you pull it in, I can merge.

**Comment by ben-allen** [comment]
@anba Rebased again! It's still at https://github.com/ben-allen/ecma402/tree/anba-826-working-branch

(also, apologies for letting this one sit so long)

File: spec/datetimeformat.html
Comment (gibson042) [inline spec/datetimeformat.html:526]:
```suggestion
              <td>Each Record represents a part of the range pattern.</td>
```

File: spec/datetimeformat.html
Comment (gibson042) [inline spec/datetimeformat.html:632]:
```suggestion
            <tr>
              <td>[[full]]</td>
              <td>a DateTime Format Record</td>
              <td>Format Record for the *"full"* style.</td>
            </tr>
            <tr>
              <td>[[long]]</td>
              <td>a DateTime Format Record</td>
              <td>Format Record for the *"long"* style.</td>
            </tr>
            <tr>
              <td>[[medium]]</td>
              <td>a DateTime Format Record</td>
              <td>Format Record for the *"medium"* style.</td>
            </tr>
            <tr>
              <td>[[short]]</td>
              <td>a DateTime Format Record</td>
              <td>Format Record for the *"short"* style.</td>
            </tr>
```

Comment (ben-allen) [reply spec/datetimeformat.html:632]: Rebased again!  Most of the comments above are addressed by #887 -- I think the only one that isn't is the use of `id` in the DateTimeFormat Patterns Record `dfn`

@anba @gibson042

**Comment by ben-allen** [comment]
I've merged the rebased version of this PR and #827 via in the new #893.

**Pull request closed without merging**
